src/quera_ahs_utils/plotting.py

```diff
@@ -248,3 +244,182 @@ def plot_avg_density(densities, register, with_labels = True, custom_axes = None
+    vertex_edge - Draw an edge around each vertex.
+    '''
+    
+    positions = np.array(positions)
+    N = len(positions)
+    
+    
+    
+    
+    
+    
+    # Construct the unit disk graph
+    G = nx.Graph((np.sqrt((positions[:,0:1] - positions[:,0:1].T)**2 + (positions[:,1::] - positions[:,1::].T)**2) + np.diag(np.ones(N))*1e99)<(2*radii))
+    # Relabel vertices
+    relabeling = {i:tuple(positions[i]) for i in range(N)}
+    G = nx.relabel.relabel_nodes(G, relabeling)
+    
+    pos = {tuple(a):a for a in positions}
+    
+    if hasattr(color_nodes,'__len__'):
+        color_nodes = {tuple(positions[i]):color_nodes[i] for i in range(N)}
+    
+    
+    '''
+    ---------------------
+    '''
+    X = positions
+    position_extent = X.max(0) - X.min(0)
+    
+    
+    # If no axis given, generate a new plot. The size will be chosen such that each
+    #  edge is a consistent size across plots.
+    if ax==None:
+        figsize = position_extent / radii + 2
+    
+        scaling = 0.5**np.ceil(np.log2(max(figsize)/16))
+        figsize *= scaling
+        
+        if scaling!=1:
+            print('Figure scaled by 1/{:0.2f}x'.format(1/scaling))
+        
+        fig = plt.figure(figsize=figsize)
+        ax = plt.subplot(1,1,1)
+        fig.subplots_adjust(left=0,right=1,top=1,bottom=0)
+    else:
+        figsize = position_extent / radii + 2
+        scaling = 0.5**np.ceil(np.log2(max(figsize)/16))
+    
+    # Determine scaling.
+    ax.axis([X.min(0)[0] - radii,X.max(0)[0] + radii,X.min(0)[1] - radii,X.max(0)[1] + radii])
+    ax.set_aspect('equal')
+    ax.axis('off')
+    
+    nodesize = 2.5*400*scaling * scale
+    edgesize = 4*scaling    * scale
+    
+    
+    
+    # Define vertex colors from a pallite
+    VERTEX_COLOR_LOOKUP = {0:'k',2:"#FFE5E5",3:"#00A2FF",1:"#C2477F"}
+    VERTEX_COLOR_LOOKUP = {0:'#333333',2:"#FF505D",3:"#6437FF",1:"#C2477F"}
+    
+    # Define edge styles from a pallite
+    EDGE_WIDTH_LOOKUP = {0:3  ,1:1     ,2:3        ,-1:0}
+    EDGE_COLOR_LOOKUP = {0:'k',1:'grey',2:'#FFE5E5',-1:'white'}
+    EDGE_COLOR_LOOKUP = {0:'#333333',1:'grey',2:'#E6E6E6',-1:'white'}
+    
+    # Clean the node color list
+    if type(color_nodes)==type(1):
+        color_nodes2 = VERTEX_COLOR_LOOKUP[color_nodes]
+    
+    elif type(color_nodes)==type('string'):
+        color_nodes2 = color_nodes
+    elif len(color_nodes)>=len(G.nodes):
+        color_nodes2 = []
+        for node in G.nodes:
+            color_nodes2.append(VERTEX_COLOR_LOOKUP.get(color_nodes[node],color_nodes[node]))
+    else:
+        print("WARNING: Color Nodes assignment not understood! Defaulting to black.")
+        color_nodes2 = 'k'
+    
+    
+    
+    # Clean the edge color list
+    if type(color_edges)==type(1):
+        color_edges2 = EDGE_COLOR_LOOKUP[color_edges]
+        edge_weights = EDGE_WIDTH_LOOKUP[color_edges]
+    elif type(color_edges)==type(('string',1)):
```

Comment:
While this DOES check if color_edges is a tuple, it doesn't verify that the contents are in the right format (a `str` followed by an `int`). 

You can take advantage of Structural Pattern Matching (SPM) in Python 3.10 by doing something like this:
```
match color_edges:
    case (str(a), int(b)):
        color_edges2 = a
        edge_weights = b
    case _:
        print("Invalid edge coloring format")
```